src/esphome/climatedevice/climatedevice.h

```diff
@@ -0,0 +1,232 @@
+#ifndef ESPHOME_CLIMATEDEVICE_CLIMATEDEVICE_H
+#define ESPHOME_CLIMATEDEVICE_CLIMATEDEVICE_H
+
+#include "esphome/defines.h"
+
+#ifdef USE_CLIMATEDEVICE
+
+#include "esphome/component.h"
+#include "esphome/automation.h"
+#include "esphome/esppreferences.h"
+#include "esphome/climatedevice/climatedevice_traits.h"
+
+ESPHOME_NAMESPACE_BEGIN
+
+namespace climatedevice {
+
+/// Simple enum to represent the mode of a climate device.
+enum ClimateDeviceMode {
+  CLIMATEDEVICE_MODE_OFF = 0,   ///< The climate device is off.
+  CLIMATEDEVICE_MODE_AUTO = 1,  ///< The climate device is in auto mode.
+  CLIMATEDEVICE_MODE_COOL = 2,  ///< The climate device is in cool mode.
+  CLIMATEDEVICE_MODE_HEAT = 3,  ///< The climate device is in heat mode.
+};
+
+template<typename... Ts> class ControlAction;
+template<typename... Ts> class PublishAction;
+template<typename... Ts> class CurrentTemperatureAction;
+
+struct ClimateDeviceState {
+  /// The mode of the climate device.
+  ClimateDeviceMode mode{CLIMATEDEVICE_MODE_OFF};
+  /// The target temperature.
+  float target_temperature{NAN};
+};
+
+#ifdef USE_MQTT_CLIMATEDEVICE
+class MQTTClimateDeviceComponent;
+#endif
+
+/** This class is climate device base class between the hardware backend and the MQTT frontend.
+ *
+ * A climate device has several variables that determine the current state: mode
+ * (OFF, AUTO, HEAT, COOL), current_temperature, target_temperature and traits (what features are supported).
+ * Both the frontend and the backend can register callbacks whenever a state is changed from the
+ * frontend and whenever a state is actually changed and should be pushed to the frontend
+ */
+class ClimateDevice : public Nameable, public PollingComponent {
```

Comment:
PollingComponent to publish the state and current temperature regularly. Since the sample rate of current_temperature may be high, the update_interval can throttle the publishing.